File: RepoTransBench/target_projects/Python/C++/softvar_json2html/public_tests/test_public_json2html_core.cpp
#include <gtest/gtest.h>
#include "json2html.h"
#include <string>
#include <map>
#include <vector>

using namespace json2html;

TEST(PublicCore, ConvertSimpleDict) {
    std::map<std::string, std::string> data = {{"animal", "Elephant"}, {"region", "Africa"}};
    std::string html = convert(data);
    EXPECT_NE(html.find("<th>animal</th>"), std::string::npos);
    EXPECT_NE(html.find("<td>Elephant</td>"), std::string::npos);
    EXPECT_NE(html.find("<th>region</th>"), std::string::npos);
    EXPECT_NE(html.find("<td>Africa</td>"), std::string::npos);
}

TEST(PublicCore, ConvertListOfNumbers) {
    std::vector<int> data = {400, 500, 600};
    std::string html = convert(data);
    EXPECT_EQ(html.find("<ul>"), 0);
    EXPECT_NE(html.find("<li>400</li>"), std::string::npos);
    EXPECT_NE(html.find("<li>500</li>"), std::string::npos);
    EXPECT_NE(html.find("<li>600</li>"), std::string::npos);
}

TEST(PublicCore, ConvertNestedDictList) {
    std::map<std::string, std::vector<std::map<std::string, std::string>>> data = {
        {"cars", { {{"make", "Toyota"}, {"year", "2010"}}, {{"make", "Ford"}, {"year", "2015"}} } }
    };
    // Flattened: just convert inside manually (simulate)
    std::string html = convert(data["cars"]);
    EXPECT_NE(html.find("Toyota"), std::string::npos);
    EXPECT_NE(html.find("Ford"), std::string::npos);
    EXPECT_TRUE(html.find("year") != std::string::npos && html.find("make") != std::string::npos);
}

TEST(PublicCore, ConvertJsonString) {
    std::string json_str = "{\"genre\": \"Jazz\", \"artist\": \"Miles Davis\"}";
    std::string html = convert(json_str);
    EXPECT_NE(html.find("<th>genre</th>"), std::string::npos);
    EXPECT_NE(html.find("<td>Jazz</td>"), std::string::npos);
    EXPECT_NE(html.find("Miles Davis"), std::string::npos);
}

TEST(PublicCore, ConvertEscapeHtml) {
    std::map<std::string, std::string> data = { {"malicious", "<img src='evil'>"} };
    std::string html = convert(data);
    EXPECT_NE(html.find("&lt;img"), std::string::npos);
}

TEST(PublicCore, ConvertBadJsonString) {
    std::string bad_json = "{\"foo\": bar";
    std::string html = convert(bad_json);
    EXPECT_TRUE(html.length() > 0);
    EXPECT_NE(html.find("{&quot;foo&quot;: bar"), std::string::npos);
}

TEST(PublicCore, ConvertEmptyObject) {
    std::map<std::string, std::string> data;
    std::string html = convert(data);
    EXPECT_EQ(html, "");
}

TEST(PublicCore, ConvertEmptyList) {
    std::vector<std::string> data;
    std::string html = convert(data);
    EXPECT_EQ(html, "");
}

File: RepoTransBench/target_projects/Python/Rust/softvar_json2html/public_tests/test_public_json2html_core.rs
use softvar_json2html::*;
use serde_json::json;

#[test]
fn test_public_convert_simple_dict() {
    let data = json!({"animal": "Elephant", "region": "Africa"});
    let html = convert(&data, None, false, true, true);
    assert!(html.contains("<th>animal</th>"));
    assert!(html.contains("<td>Elephant</td>"));
    assert!(html.contains("<th>region</th>"));
    assert!(html.contains("<td>Africa</td>"));
}

#[test]
fn test_public_convert_list_of_numbers() {
    let data = json!([400, 500, 600]);
    let html = convert(&data, None, false, true, true);
    assert!(html.starts_with("<ul>"));
    assert!(html.contains("<li>400</li>"));
    assert!(html.contains("<li>500</li>"));
    assert!(html.contains("<li>600</li>"));
}

#[test]
fn test_public_convert_nested_dict_list() {
    let data = json!({"cars": [
        {"make": "Toyota", "year": 2010},
        {"make": "Ford", "year": 2015}
    ]});
    let html = convert(&data, None, false, true, true);
    assert!(html.contains("Toyota"));
    assert!(html.contains("Ford"));
    assert!(html.contains("year"));
    assert!(html.contains("make"));
}

#[test]
fn test_public_convert_json_string() {
    let json_str = r#"{"genre": "Jazz", "artist": "Miles Davis"}"#;
    let html = convert(&json!(json_str), None, false, true, true);
    assert!(html.contains("<th>genre</th>"));
    assert!(html.contains("<td>Jazz</td>"));
    assert!(html.contains("Miles Davis"));
}

#[test]
fn test_public_convert_escape_html() {
    let data = json!({"malicious": "<img src='evil'>"});
    let html = convert(&data, None, false, true, true);
    assert!(html.contains("&lt;img"));
}

#[test]
fn test_public_convert_bad_json_string() {
    let bad_json = r#"{"foo": bar"#;
    let html = convert(&json!(bad_json), None, false, true, true);
    assert!(html.contains("{&quot;foo&quot;: bar") || html.contains(bad_json));
}

#[test]
fn test_public_convert_empty_object() {
    let data = json!({});
    let html = convert(&data, None, false, true, true);
    assert_eq!(html, "");
}

#[test]
fn test_public_convert_empty_list() {
    let data = json!([]);
    let html = convert(&data, None, false, true, true);
    assert_eq!(html, "");
}